imports/server/MigrationRegistry.ts

```diff
@@ -0,0 +1,200 @@
+// * We update the recorded version after each successful migration, to avoid
+//   the following scenario:
+//   * backend A tries to migrate to version 2, sees DB at version 0
+//   * backend A locks control
+//   * backend A runs migration 1
+//   * backend A runs migration 2 but it crashes
+//   * backend B comes along and migration 1 has fully applied, but if control
+//     is manually unlocked it'll get run again.
+
+import { Mongo } from 'meteor/mongo';
+import ignoringDuplicateKeyErrors from './ignoringDuplicateKeyErrors';
+
+export type Migration = {
+  version: number;
+  name: string;
+  up: () => void;
+}
+
+export type MigrationControl = {
+  _id: string;
+  version: number;
+  locked: boolean;
+  lockedAt?: Date;
+}
+
+class MigrationRegistry {
+  private collection: Mongo.Collection<MigrationControl>;
+
+  private migrations: Migration[];
+
+  private shouldLog: boolean;
+
+  constructor({ collection }: {
+    collection?: Mongo.Collection<MigrationControl>;
+  }) {
+    this.collection = collection ?? new Mongo.Collection<MigrationControl>('migrations');
+    this.migrations = [];
+    this.shouldLog = true;
+  }
+
+  config({ log }: {
+    log: boolean;
+  }) {
+    this.shouldLog = log;
+  }
+
+  log(...args: any[]): void {
+    if (this.shouldLog) {
+      // eslint-disable-next-line no-console
+      console.log(...args);
+    }
+  }
+
+  add(migration: Migration): void {
+    // Registers a migration.  This should first be called on a Migration with version 1,
+    // then second on a migration with version 2, and so on, keeping the registered versions
+    // sequential and dense.
+
+    // Require migration versions to be ordered, linear, and dense.
+    if (migration.version !== (this.migrations.length + 1)) {
+      throw new Error(`Migrations must be registered in sequential order \
+        (expected version ${this.migrations.length}, but "${migration.name}" had version ${migration.version}`);
+    }
+    this.migrations.push(migration);
+  }
+
+  migrateToLatest(): boolean {
+    return this.migrateTo(this.migrations.length);
+  }
+
+  migrateTo(target: number): boolean {
+    // Returns true if we believe we have successfully reached the target
+    // version; false if control record was locked or the observed version
+    // exceeded our target version.
+    const observedVersion = this.getVersion();
+    if (observedVersion === target) {
+      this.log(`Migrations: Not migrating, already at version ${observedVersion}`);
+      return true;
+    } else if (observedVersion > target) {
+      this.log(`Migrations: not migrating, observed version ${observedVersion} exceeds target ${target}`);
+      return false;
+    }
+    // by elimination, observedVersion < target.  We have migrations to run.
+    // Acquire the lock.  We might race with another backend that is also
+    // trying to apply this same migration.
+    if (!this.lock()) {
+      this.log('Not migrating, control is locked');
+      return false;
```

Comment:
If we're optimizing for not requiring administrative intervention, is it worth having a prevention mechanism in case the lock was acquired in the distant past? (Say, more than an hour ago)

Reply:
Yeah I don't think Mongo has anything of that shape. Historically, Mongo has shied away from semantics that are tied to connection state because it makes it more difficult to behave correctly in the event of a failover. I think there's a bit more these days (especially around the causal consistency features) but still nothing like that.

Reply:
Updated behavior here to preempt if the lock is >10 minutes old (and added a test that preemption works as expected).